### ISSUE TITLE
weight_norm doesn't support eta and returns nan for zero weights

### ISSUE THREAD

**Issue by fuzic** [pytorch/pytorch#19126]
## 🐛 Bug

backprop on weights generated with torch._weight_norm that are zero filled yields nan gradients. I don't see a way to add an eta to the norm to prevent this.

## To Reproduce

```
w = torch.zeros(1, 3)
v = w / w.norm()
g = w.norm()
v.requires_grad = True
g.requires_grad = True
torch._weight_norm(v, g).matmul(torch.randn(3,1)).backward()
v.grad
> tensor([[nan, nan, nan]])
g.grad
> tensor(nan)
```

I'm encountering nan's during backprop during training of a network with weight normalization. From this seemingly related thread it sounds like the advice is to add an eta to the norm, but in this case the norm is generated in pytorch's c++ implementation and I don't see an obvious way to do this.

## Expected behavior

I expect there to be a way to generate non-nan gradients for weight-norm weights that are zero filled

## Environment

 - PyTorch Version (e.g., 1.0): 1.0.0
 - OS (e.g., Linux): Mac and Centos
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.6
 - CUDA/cuDNN version: None
 - GPU models and configuration: None
 - Any other relevant information:

**Comment by SsnL** [comment]
you are using non public API and the gradient comes from the fact that you didn't add eps/eta in denominator yourself. use this instead: https://pytorch.org/docs/master/nn.html#torch.nn.utils.weight_norm

**Comment by fuzic** [comment]
thanks @SsnL, I can reproduce something similar with the public weight_norm api. How do I add an eps/eta in the denominator myself?

>>> from torch.nn.utils import weight_norm
>>> import torch.nn as nn
>>> import torch
>>> l = nn.Conv1d(3, 1, 1)
>>> l.weight.data.normal_(0, 1e-100)
tensor([[[0.],
         [0.],
         [0.]]])
>>> wl = weight_norm(l)
>>> o = wl(torch.zeros(1, 3, 1))
>>> o.backward()
>>> [p.grad for p in wl.parameters()]
[tensor([1.]), tensor([[[nan]]]), tensor([[[nan],
         [nan],
         [nan]]])]

**Comment by shreydesai** [comment]
I'm having this problem as well. Any update?

**Comment by shreydesai** [comment]
Following up

**Comment by umgupta** [comment]
I think I encounter this too. Looking at the implementation https://pytorch.org/docs/master/_modules/torch/nn/utils/weight_norm.html#weight_norm

I think modifying the `compute_weight` function as below should do the trick.

```
def compute_weight(self, module):

        g = getattr(module, self.name + '_g')
        v = getattr(module, self.name + '_v')
        w = v*(g/(torch.norm(v)+1e-8)).expand_as(v)

        return w
```
Inspired by https://gist.github.com/rtqichen/b22a9c6bfc4f36e605a7b3ac1ab4122f

**Comment by rvaghefi** [comment]
is there any update on this issue? the suggested solution by @umgupta degrades my validation loss even when I set 1e-8 to zero. I guess norm_except_dim and _weight_norm have different implementation in compute_weight.

**Comment by mruberry** [comment]
No update, sorry @rvaghefi. While we expect to work on reparameterization in the future it hasn't come up yet.

**Comment by asyounis** [comment]
Any updates on this issue?  Maybe we can add a eta parameter as input with a default set to 1e-8?
